src/GafferSceneModule/AttributesBinding.cpp

```diff
@@ -53,10 +54,84 @@
 #include "GafferScene/StandardAttributes.h"
 
 #include "GafferBindings/DependencyNodeBinding.h"
+#include "GafferBindings/Serialisation.h"
+
+#include <cassert>
 
 using namespace boost::python;
 using namespace GafferScene;
 
+namespace AttributeQueryBinding
+{
+
+bool isSetup( AttributeQuery const& query )
+{
+	IECorePython::ScopedGILRelease gilRelease;
+	return query.isSetup();
+}
+
+bool canSetup( AttributeQuery const& query, Gaffer::Plug const& plug )
+{
+	IECorePython::ScopedGILRelease gilRelease;
+	return query.canSetup( & plug );
+}
+
+void setup( AttributeQuery& query, Gaffer::Plug const& plug )
+{
+	IECorePython::ScopedGILRelease gilRelease;
+	if( !( query.canSetup( & plug ) ) )
+	{
+		throw IECore::Exception( "AttributeQuery cannot be setup from specified plug" );
+	}
+	query.setup( & plug );
+}
+
+class Serialiser : public GafferBindings::NodeSerialiser
+{
+	bool childNeedsConstruction( Gaffer::GraphComponent const* child, GafferBindings::Serialisation const& serialisation ) const override;
+	std::string postConstructor( Gaffer::GraphComponent const* component, std::string const& identifier, GafferBindings::Serialisation& serialisation ) const override;
+};
+
+bool Serialiser::childNeedsConstruction( Gaffer::GraphComponent const* const child, GafferBindings::Serialisation const& serialisation ) const
+{
+	AttributeQuery const* const query = child->parent< AttributeQuery >();
+	assert( query != 0 );
+
+	return ( !(
+			( child == query->defaultPlug() ) ||
+			( child == query->valuePlug() ) ) ) &&
+		( GafferBindings::NodeSerialiser::childNeedsConstruction( child, serialisation ) );
+}
+
+std::string Serialiser::postConstructor( Gaffer::GraphComponent const* const component, std::string const& identifier, GafferBindings::Serialisation& serialisation ) const
+{
+	std::string result = GafferBindings::NodeSerialiser::postConstructor( component, identifier, serialisation );
+
+	AttributeQuery const* const query = IECore::assertedStaticCast< AttributeQuery const >( component );
+
+	if( query->isSetup() )
+	{
+		if( result.size() )
+		{
+			result += "\n";
+		}
+
+		Gaffer::PlugPtr plug = query->valuePlug()->createCounterpart( "in", Gaffer::Plug::In );
+		assert( plug );
+
+		plug->setFlags( Gaffer::Plug::Dynamic, false );
```

Comment:
The plug is being used to generate a serialised call to the plug constructor so that a plug of the correct type can be passed to setup() on load. This is the same pattern used by the other setup() nodes. I was curious why the setFlags() call was necessary.

Reply:
Might be easier to explain with code - this is how I think we should implement it : https://github.com/johnhaddon/gaffer/commit/6c3afba98d115976b5fbb9362432f49c1ba5e9ac. The examples you've started from have a bunch of historical cruft that I'm slowly trying to refactor out.